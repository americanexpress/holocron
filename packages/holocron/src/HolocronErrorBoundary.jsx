import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import hoistNonReactStatics from 'hoist-non-react-statics';

class HolocronErrorBoundary extends Component {
  constructor(props) {
    super(props);
    this.state = { error: null };
  }

  static getDerivedStateFromError(error) {
    return { error };
  }

  // eslint-disable-next-line class-methods-use-this
  componentDidCatch(error, errorInfo) {
    console.error(error);
    console.error(errorInfo);
  }

  render() {
    const { props: { children }, state: { error } } = this;

    if (error) {
      if (process.env.NODE_ENV === 'production') return null;
      return <div style={{ color: 'red' }}>{error.message}</div>;
    }

    return (
      <Fragment>
        {children}
      </Fragment>
    );
  }
}

HolocronErrorBoundary.propTypes = {
  children: PropTypes.node,
};

HolocronErrorBoundary.defaultProps = {
  children: null,
};

export const withHolocronErrorBoundary = (WrappedComponent) => {
  const name = WrappedComponent.displayName || WrappedComponent.name;
  const Wrapped = props => (
    <HolocronErrorBoundary>
      <WrappedComponent {...props} />
    </HolocronErrorBoundary>
  );

  Wrapped.displayName = name
    ? `WithHolocronErrorBoundary(${name})`
    : 'WithHolocronErrorBoundary';

  return hoistNonReactStatics(Wrapped, WrappedComponent);
};

export default HolocronErrorBoundary;
