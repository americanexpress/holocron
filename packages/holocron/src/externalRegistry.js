/*
 * Copyright 2023 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import { fromJS, List } from 'immutable';

let externals = fromJS({});
let externalMap = fromJS({});
let externalBlockList = List();

function addToExternalBlockList(externalUrl) {
  externalBlockList = externalBlockList.push(externalUrl);
}

function isExternalInBlockList(externalUrl) {
  return externalBlockList.find((value) => value === externalUrl);
}

function getExternalBlockList() {
  return externalBlockList;
}

// function registerExternal(externalName, external) {
//   externals = externals.set(externalName, addHigherOrderComponent(external));
// }

function getExternal(externalName, altExternals) {
  if (altExternals) {
    return altExternals.get(externalName);
  }

  return externals.get(externalName);
}

function getExternals() {
  return externals;
}

function getExternalMap() {
  return externalMap;
}

function setExternalMap(newExternalMap) {
  externalMap = fromJS(newExternalMap);
}

function resetExternalRegistry(newExternals, newExternalMap) {
  externals = fromJS(newExternals);
  setExternalMap(newExternalMap);
}

export {
  addToExternalBlockList,
  isExternalInBlockList,
  getExternalBlockList,
  // registerExternal,
  getExternal,
  getExternals,
  getExternalMap,
  setExternalMap,
  resetExternalRegistry,
};
