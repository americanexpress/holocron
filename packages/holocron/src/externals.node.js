import { fromJS } from 'immutable';
import requireFromString from 'require-from-string';

let sharedExternals = fromJS({});

function registerSharedExternal(externalName, external) {
  sharedExternals = sharedExternals.set(externalName, external);
}

async function loadSharedExternal(externalName, url) {
  try {
    const externalResp = await fetch(url);
    const externalString = await externalResp.text();
    const external = requireFromString(externalString, url);
    registerSharedExternal(externalName, external);
    return external;
  } catch (e) {
    console.log(`Error fetching External ${externalName}`, e);
    return null;
  }
}

function getExternal(externalName) {
  if (sharedExternals.get(externalName)) {
    return sharedExternals.get(externalName);
  }
  throw new Error(`${externalName} does not exist!`);
}

export {
  registerSharedExternal,
  loadSharedExternal,
  getExternal,
};
