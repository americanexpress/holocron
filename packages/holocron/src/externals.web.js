function getExternal(externalName) {
  const formattedExternalName = externalName.replace(/-/g, '_');
  const globalExternalVar = `__holocron_external_${formattedExternalName}`;
  if (window[globalExternalVar]) {
    return window[globalExternalVar];
  }
  throw new Error(`Unable to find external ${externalName}`);
}

export default getExternal;
