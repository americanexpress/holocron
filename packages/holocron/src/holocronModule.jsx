/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createSelector } from 'reselect';
import hoistStatics from 'hoist-non-react-statics';

import {
  LOAD_KEY,
  REDUCER_KEY,
  MODULES_STORE_KEY,
  INIT_MODULE_STATE,
} from './ducks/constants';

// Execute deprecated load function and provide deprecation message
export function executeLoad({ dispatch, load, ...restProps } = { }) {
  if (load) {
    console.warn('The \'load\' function in holocron has been deprecated. Please use \'loadModuleData\' instead.');
    return dispatch(load(restProps));
  }
  return undefined;
}

// Dispatch loadModuleData inside a thunk if it exists
export function executeLoadModuleData(loadModuleData, WrappedComponent, props) {
  const { dispatch, ...restProps } = props;
  if (loadModuleData) {
    return dispatch(async (_, getState, { fetchClient }) => loadModuleData({
      store: { dispatch, getState },
      fetchClient,
      ownProps: restProps,
      module: WrappedComponent,
    }));
  }
  return undefined;
}

export function getModuleName(WrappedComponent, name) {
  return WrappedComponent && (WrappedComponent.displayName || WrappedComponent.name || name);
}

export function getModuleDisplayName(name) {
  return `HolocronModule(${name})`;
}

export async function executeLoadingFunctions({
  // Provide loadModuleData function
  loadModuleData,
  // Provide WrappedComponent for loadModuleData
  WrappedComponent,
  // Frozen props as of when called
  frozenProps,
  // Provide loadCount to limit state changes
  loadCount,
  // Provide componentName for error messages
  componentName,
  // Provide component instance for getting current state of component post async
  hocInstance,
}) {
  try {
    // Call deprecated load function if exists
    await executeLoad(frozenProps);
    // Call loadModuleData if it exists
    await executeLoadModuleData(loadModuleData, WrappedComponent, frozenProps);
    // Modify state only when mounted and current loadCount is less or equal than previous loadCount
    if (hocInstance.mounted && hocInstance.state.loadCount <= loadCount) {
      hocInstance.setState({ status: 'loaded' });
    }
  } catch (error) {
    console.error(`Error while attempting to call 'load' or 'loadModuleData' inside Holocron module ${componentName}.`, error);
    if (hocInstance.mounted) {
      hocInstance.setState({ status: 'error' });
    }
  }
}

export default function holocronModule({
  name,
  reducer,
  load, // TODO remove in next major version
  shouldModuleReload,
  loadModuleData,
  mergeProps,
  options = {},
} = {}) {
  return function wrapWithHolocron(WrappedComponent) {
    class HolocronModuleWrapper extends React.Component {
      constructor(props) {
        super(props);
        this.state = {
          loadCount: 0,
          status: 'loading',
        };
        this.mounted = false;
      }

      componentDidMount() {
        this.mounted = true;
        // eslint-disable-next-line no-underscore-dangle
        if ((loadModuleData || load) && !global.__INITIAL_STATE__) {
          this.initiateLoad(0, this.props);
        }
      }

      // ignoring to support deprecated componentWillReceiveProps.
      // This needs to be removed to support React17
      // eslint-disable-next-line camelcase
      UNSAFE_componentWillReceiveProps(nextProps) {
        const { loadCount } = this.state;
        const { props } = this;
        if (shouldModuleReload && shouldModuleReload(props, nextProps)) {
          const newLoadCount = loadCount + 1;
          this.setState({ status: 'loading', loadCount: newLoadCount });
          this.initiateLoad(newLoadCount, nextProps);
        }
      }

      // Ignoring as internal state cannot be checked when unmounted.
      /* istanbul ignore next */
      componentWillUnmount() {
        this.mounted = false;
      }

      initiateLoad(loadCount, frozenProps) {
        return executeLoadingFunctions({
          loadModuleData,
          WrappedComponent,
          frozenProps,
          loadCount,
          hocName: getModuleName(WrappedComponent, name),
          hocInstance: this,
        });
      }

      render() {
        const { status } = this.state;
        // eslint-disable-next-line react/jsx-props-no-spreading
        return <WrappedComponent {...this.props} moduleLoadStatus={status} />;
      }
    }

    HolocronModuleWrapper.propTypes = {
      load: PropTypes.func,
      dispatch: PropTypes.func.isRequired,
    };

    HolocronModuleWrapper.defaultProps = {
      load: undefined,
    };

    HolocronModuleWrapper.displayName = getModuleDisplayName(getModuleName(WrappedComponent, name));

    if (load && (options.ssr || global.BROWSER)) {
      HolocronModuleWrapper[LOAD_KEY] = load;
    }

    let mapModuleStateToProps;

    if (reducer && !name) {
      console.warn(`The Holocron Config in '${getModuleDisplayName(getModuleName(WrappedComponent, name))}' requires a 'name' when passing a 'reducer'.\nThe 'reducer' will not be added to the Redux Store without a 'name'.`);
    }

    if (reducer && name) {
      HolocronModuleWrapper[REDUCER_KEY] = reducer;
      const getModuleState = createSelector(
        (state) => state.getIn(
          [MODULES_STORE_KEY, name],
          reducer(undefined, { type: INIT_MODULE_STATE })
        ),
        (moduleState) => moduleState.toJS()
      );

      mapModuleStateToProps = ((state) => {
        const moduleState = getModuleState(state);
        return { moduleState };
      });
    }

    const mapDispatchToProps = (dispatch) => {
      if (load) {
        return {
          load,
          dispatch,
        };
      }
      return { dispatch };
    };

    hoistStatics(HolocronModuleWrapper, WrappedComponent);

    const DerivedComponent = connect(
      mapModuleStateToProps,
      mapDispatchToProps,
      mergeProps
    )(HolocronModuleWrapper);

    hoistStatics(DerivedComponent, HolocronModuleWrapper);

    return DerivedComponent;
  };
}
