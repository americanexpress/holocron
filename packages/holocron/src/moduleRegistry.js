/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import { fromJS, List } from 'immutable';
import holocronModule from './holocronModule';

let modules = fromJS({});
let moduleMap = fromJS({});
let moduleBlockList = List();

function addToModuleBlockList(moduleUrl) {
  moduleBlockList = moduleBlockList.push(moduleUrl);
}

function isModuleInBlockList(moduleUrl) {
  return moduleBlockList.find((value) => value === moduleUrl);
}

function getModuleBlockList() {
  return moduleBlockList;
}

function addHigherOrderComponent(module) {
  return holocronModule({ loadModuleData: module.loadModuleData, ...module.holocron })(module);
}

function registerModule(moduleName, module) {
  modules = modules.set(moduleName, addHigherOrderComponent(module));
}

function getModule(moduleName, altModules) {
  if (altModules) {
    return altModules.get(moduleName);
  }

  return modules.get(moduleName);
}

function getModules() {
  return modules;
}

function getModuleMap() {
  return moduleMap;
}

function setModuleMap(newModuleMap) {
  moduleMap = fromJS(newModuleMap);
}

function resetModuleRegistry(newModules, newModuleMap) {
  modules = fromJS(newModules);
  setModuleMap(newModuleMap);
}

export {
  addToModuleBlockList,
  getModuleBlockList,
  isModuleInBlockList,
  registerModule,
  getModule,
  getModules,
  getModuleMap,
  setModuleMap,
  resetModuleRegistry,
  addHigherOrderComponent,
};
