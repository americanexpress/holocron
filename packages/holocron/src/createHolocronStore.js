/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import { createStore, compose } from 'redux';
import immutableCombineReducers from '@americanexpress/vitruvius/immutable';
import thunk from 'redux-thunk';
import { Map as iMap, Set as iSet, isImmutable } from 'immutable';

import holocronReducer from './ducks/load';
import { getModule, getModules } from './moduleRegistry';
import {
  HOLOCRON_STORE_KEY, MODULES_STORE_KEY, REDUCER_KEY, MODULE_REDUCER_ADDED,
} from './ducks/constants';

function immutableCombineReducersWithNewModules(moduleReducerMap, newModuleState) {
  if (Object.keys(moduleReducerMap).length === 0) {
    const emptyState = iMap();
    return () => emptyState;
  }

  const baseReducer = immutableCombineReducers(moduleReducerMap);

  if (Object.keys(newModuleState).length === 0) {
    return baseReducer;
  }

  return (state, action) => {
    if (action.type === MODULE_REDUCER_ADDED) {
      return state.merge(newModuleState);
    }
    return baseReducer(state, action);
  };
}

function createReducer({ appReducer, state, localsForBuildInitialState }) {
  const moduleReducerMap = {};
  const newModuleState = {};

  let withReducers = iSet();
  let loadedReducers = iMap();

  if (state) {
    withReducers = state.getIn([HOLOCRON_STORE_KEY, 'withReducers'], withReducers);
    loadedReducers = state.get(MODULES_STORE_KEY, loadedReducers);
  }

  withReducers.forEach((moduleName) => {
    const module = getModule(moduleName);
    if (!module) {
      console.warn(`unable to get the reducer of holocron module ${moduleName}`);
      return;
    }
    const moduleReducer = module[REDUCER_KEY];
    if (typeof moduleReducer.buildInitialState === 'function' && !loadedReducers.has(moduleName)) {
      newModuleState[moduleName] = moduleReducer.buildInitialState(
        localsForBuildInitialState
      );
    }
    moduleReducerMap[moduleName] = moduleReducer;
  });

  const holocronReducers = immutableCombineReducers({
    [HOLOCRON_STORE_KEY]: holocronReducer,
    [MODULES_STORE_KEY]: immutableCombineReducersWithNewModules(moduleReducerMap, newModuleState),
  });

  const holocronStoreKeys = [HOLOCRON_STORE_KEY, MODULES_STORE_KEY];

  return (storeState = iMap(), action) => {
    const holocronState = holocronReducers(
      storeState.filter((_, key) => holocronStoreKeys.includes(key)), action);
    const appState = appReducer(
      storeState.filter((_, key) => !holocronStoreKeys.includes(key)), action);

    return appState.merge(holocronState);
  };
}

const holocronEnhancer = (localsForBuildInitialState, extraThunkArguments = {}) => (
  createReduxStore
) => (
  appReducer,
  preloadedState,
  enhancer
) => {
  const reducer = createReducer({
    appReducer,
    state: preloadedState,
    localsForBuildInitialState,
  });
  const store = createReduxStore(reducer, preloadedState, enhancer);
  const { getState } = store;
  const rebuildReducer = () => store.replaceReducer(createReducer({
    appReducer,
    state: getState(),
    localsForBuildInitialState,
  }));
  store.rebuildReducer = rebuildReducer;
  let { dispatch } = store;

  if (!global.BROWSER) { store.modules = getModules(); }

  const middlewareAPI = {
    getState: store.getState,
    dispatch: (action) => dispatch(action),
  };
  dispatch = thunk.withExtraArgument({
    ...extraThunkArguments,
    rebuildReducer,
    modules: store.modules,
  })(middlewareAPI)(store.dispatch);

  return {
    ...store,
    dispatch,
  };
};

export default function createHolocronStore({
  reducer,
  initialState,
  enhancer,
  localsForBuildInitialState,
  extraThunkArguments,
}) {
  if (!(!initialState || isImmutable(initialState))) {
    throw new Error('createHolocronStore expects immutable initial state');
  }

  const enhancedEnhancer = enhancer
    ? compose(holocronEnhancer(localsForBuildInitialState, extraThunkArguments), enhancer)
    : holocronEnhancer(localsForBuildInitialState, extraThunkArguments);

  return createStore(reducer, initialState, enhancedEnhancer);
}
