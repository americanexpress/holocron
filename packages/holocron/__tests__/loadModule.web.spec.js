/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import { fromJS } from 'immutable';
import loadModule from '../src/loadModule.web';
import { resetModuleRegistry } from '../src/moduleRegistry';

let mockElement;
let map;

jest.spyOn(document, 'createElement').mockImplementation(() => mockElement);
jest.spyOn(document, 'getElementsByTagName').mockImplementation(() => [{ appendChild: () => null }]);
// eslint-disable-next-line no-underscore-dangle
window.__holocron_module_bundle_type__ = 'browser';

describe('loadModule.web', () => {
  beforeEach(() => {
    map = {};
    mockElement = {
      addEventListener: jest.fn((event, cb) => {
        map[event] = cb;
      }),
    };
  });

  it('should add crossorigin to scripts', () => {
    loadModule(
      'my-module',
      fromJS({
        node: {
          url: 'https://example.com/cdn/my-module/1.0.0/my-module.node.js',
          integrity: '123',
        },
        browser: {
          url: 'https://example.com/cdn/my-module/1.0.0/my-module.browser.js',
          integrity: '234',
        },
        legacyBrowser: {
          url: 'https://example.com/cdn/my-module/1.0.0/my-module.legacy.browser.js',
          integrity: '344',
        },
      })
    );
    expect(mockElement.crossOrigin).toBe('anonymous');
  });

  it('should add the correct integrity to scripts if NODE_ENV is production', () => {
    process.env.NODE_ENV = 'production';
    loadModule(
      'my-module',
      fromJS({
        node: {
          url: 'https://example.com/cdn/my-module/1.0.0/my-module.node.js',
          integrity: '123',
        },
        browser: {
          url: 'https://example.com/cdn/my-module/1.0.0/my-module.browser.js',
          integrity: '234',
        },
        legacyBrowser: {
          url: 'https://example.com/cdn/my-module/1.0.0/my-module.legacy.browser.js',
          integrity: '344',
        },
      })
    );
    expect(mockElement.integrity).toBe('234');
  });

  it('should not add integrity to scripts if NODE_ENV is development', () => {
    process.env.NODE_ENV = 'development';

    loadModule(
      'my-module',
      fromJS({
        node: {
          url: 'https://example.com/cdn/my-module/1.0.0/my-module.node.js',
          integrity: '123',
        },
        browser: {
          url: 'https://example.com/cdn/my-module/1.0.0/my-module.browser.js',
          integrity: '234',
        },
        legacyBrowser: {
          url: 'https://example.com/cdn/my-module/1.0.0/my-module.legacy.browser.js',
          integrity: '344',
        },
      })
    );
    expect(mockElement.integrity).toBeUndefined();
  });

  it('should reject with a helpful error when module name is not a string', async () => {
    await expect(
      loadModule(
        NaN,
        fromJS({
          node: {
            url: 'https://example.com/cdn/my-module/1.0.0/my-module.node.js',
            integrity: '123',
          },
          browser: {
            url: 'https://example.com/cdn/my-module/1.0.0/my-module.browser.js',
            integrity: '234',
          },
          legacyBrowser: {
            url: 'https://example.com/cdn/my-module/1.0.0/my-module.legacy.browser.js',
            integrity: '344',
          },
        })
      )
    ).rejects.toMatchSnapshot();
  });

  it('should reject with a helpful error when module data is not an object', async () => {
    await expect(
      loadModule(
        'my-module',
        undefined
      )
    ).rejects.toMatchSnapshot();
  });

  it('should reject with the error on error', () => {
    const loadPromise = loadModule(
      'erroring-module',
      fromJS({
        node: {
          url: 'https://example.com/cdn/erroring-module/1.0.0/erroring-module.node.js',
          integrity: '123',
        },
        browser: {
          url: 'https://example.com/cdn/erroring-module/1.0.0/erroring-module.browser.js',
          integrity: '234',
        },
        legacyBrowser: {
          url: 'https://example.com/cdn/erroring-module/1.0.0/erroring-module.legacy.browser.js',
          integrity: '344',
        },
      })
    );
    const loadError = new Error('failed to load module');
    expect(mockElement.addEventListener.mock.calls[0][0]).toBe('error');
    expect(mockElement.addEventListener.mock.calls[0][1](loadError)).toBeUndefined();
    return expect(loadPromise).rejects.toBe(loadError.message);
  });

  it('should resolve with the module on load', () => {
    const LoadingModule = () => 'hello';
    resetModuleRegistry(
      {
        'loading-module': LoadingModule,
      },
      {
        modules: {
          'loading-module': {
            node: {
              url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.node.js',
              integrity: '123',
            },
            browser: {
              url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.browser.js',
              integrity: '234',
            },
            legacyBrowser: {
              url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.legacy.browser.js',
              integrity: '344',
            },
          },
        },
      }
    );
    const loadPromise = loadModule(
      'loading-module',
      fromJS({
        node: {
          url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.node.js',
          integrity: '123',
        },
        browser: {
          url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.browser.js',
          integrity: '234',
        },
        legacyBrowser: {
          url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.legacy.browser.js',
          integrity: '344',
        },
      })
    );
    expect(mockElement.addEventListener.mock.calls[1][0]).toBe('load');
    expect(mockElement.addEventListener.mock.calls[1][1]()).toBeUndefined();
    return expect(loadPromise).resolves.toBe(LoadingModule);
  });

  it('should add the module map clientCacheRevision to the script tag src for cache busting purposes if NODE_ENV is production', async () => {
    process.env.NODE_ENV = 'production';
    const LoadingModule = () => 'hello';
    resetModuleRegistry(
      {
        'loading-module': LoadingModule,
      },
      {
        clientCacheRevision: 'key123',
        modules: {
          'loading-module': {
            node: {
              url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.node.js',
              integrity: '123',
            },
            browser: {
              url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.browser.js',
              integrity: '234',
            },
            legacyBrowser: {
              url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.legacy.browser.js',
              integrity: '344',
            },
          },
        },
      }
    );
    loadModule(
      'loading-module',
      fromJS({
        node: {
          url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.node.js',
          integrity: '123',
        },
        browser: {
          url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.browser.js',
          integrity: '234',
        },
        legacyBrowser: {
          url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.legacy.browser.js',
          integrity: '344',
        },
      })
    );
    expect(mockElement.addEventListener.mock.calls[1][0]).toBe('load');
    expect(mockElement.addEventListener.mock.calls[1][1]()).toBeUndefined();
    expect(mockElement.src).toBe('https://example.com/cdn/loading-module/1.0.0/loading-module.browser.js?clientCacheRevision=key123');
    expect(new URL(mockElement.src).search).toBe('?clientCacheRevision=key123');
  });

  it('should fallback to module map key if it is provided in place of clientCacheRevision', async () => {
    process.env.NODE_ENV = 'production';
    const LoadingModule = () => 'hello';
    resetModuleRegistry(
      {
        'loading-module': LoadingModule,
      },
      {
        key: 'key456',
        modules: {
          'loading-module': {
            node: {
              url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.node.js',
              integrity: '123',
            },
            browser: {
              url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.browser.js',
              integrity: '234',
            },
            legacyBrowser: {
              url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.legacy.browser.js',
              integrity: '344',
            },
          },
        },
      }
    );
    loadModule(
      'loading-module',
      fromJS({
        node: {
          url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.node.js',
          integrity: '123',
        },
        browser: {
          url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.browser.js',
          integrity: '234',
        },
        legacyBrowser: {
          url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.legacy.browser.js',
          integrity: '344',
        },
      })
    );
    expect(mockElement.addEventListener.mock.calls[1][0]).toBe('load');
    expect(mockElement.addEventListener.mock.calls[1][1]()).toBeUndefined();
    expect(mockElement.src).toBe('https://example.com/cdn/loading-module/1.0.0/loading-module.browser.js?clientCacheRevision=key456');
    expect(new URL(mockElement.src).search).toBe('?clientCacheRevision=key456');
  });

  it('should use clientCacheRevision if key is also provided', async () => {
    process.env.NODE_ENV = 'production';
    const LoadingModule = () => 'hello';
    resetModuleRegistry(
      {
        'loading-module': LoadingModule,
      },
      {
        key: 'key456',
        clientCacheRevision: 'abc',
        modules: {
          'loading-module': {
            node: {
              url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.node.js',
              integrity: '123',
            },
            browser: {
              url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.browser.js',
              integrity: '234',
            },
            legacyBrowser: {
              url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.legacy.browser.js',
              integrity: '344',
            },
          },
        },
      }
    );
    loadModule(
      'loading-module',
      fromJS({
        node: {
          url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.node.js',
          integrity: '123',
        },
        browser: {
          url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.browser.js',
          integrity: '234',
        },
        legacyBrowser: {
          url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.legacy.browser.js',
          integrity: '344',
        },
      })
    );
    expect(mockElement.addEventListener.mock.calls[1][0]).toBe('load');
    expect(mockElement.addEventListener.mock.calls[1][1]()).toBeUndefined();
    expect(mockElement.src).toBe('https://example.com/cdn/loading-module/1.0.0/loading-module.browser.js?clientCacheRevision=abc');
    expect(new URL(mockElement.src).search).toBe('?clientCacheRevision=abc');
  });

  it('should not add the module map clientCacheRevision to the script tag src for cache busting purposes if NODE_ENV is development', async () => {
    process.env.NODE_ENV = 'development';
    const LoadingModule = () => 'hello';
    resetModuleRegistry(
      {
        'loading-module': LoadingModule,
      },
      {
        clientCacheRevision: 'key123',
        modules: {
          'loading-module': {
            node: {
              url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.node.js',
              integrity: '123',
            },
            browser: {
              url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.browser.js',
              integrity: '234',
            },
            legacyBrowser: {
              url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.legacy.browser.js',
              integrity: '344',
            },
          },
        },
      }
    );
    loadModule(
      'loading-module',
      fromJS({
        node: {
          url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.node.js',
          integrity: '123',
        },
        browser: {
          url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.browser.js',
          integrity: '234',
        },
        legacyBrowser: {
          url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.legacy.browser.js',
          integrity: '344',
        },
      })
    );
    expect(mockElement.addEventListener.mock.calls[1][0]).toBe('load');
    expect(mockElement.addEventListener.mock.calls[1][1]()).toBeUndefined();
    expect(mockElement.src).toBe('https://example.com/cdn/loading-module/1.0.0/loading-module.browser.js');
    expect(new URL(mockElement.src).search).toBe('');
  });

  it('does not add the module map clientCacheRevision to the script tag src for cache busting purposes when no clientCacheRevision', async () => {
    process.env.NODE_ENV = 'production';
    const LoadingModule = () => 'hello';
    resetModuleRegistry(
      {
        'loading-module': LoadingModule,
      },
      {
        modules: {
          'loading-module': {
            node: {
              url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.node.js',
              integrity: '123',
            },
            browser: {
              url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.browser.js',
              integrity: '234',
            },
            legacyBrowser: {
              url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.legacy.browser.js',
              integrity: '344',
            },
          },
        },
      }
    );
    loadModule(
      'loading-module',
      fromJS({
        node: {
          url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.node.js',
          integrity: '123',
        },
        browser: {
          url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.browser.js',
          integrity: '234',
        },
        legacyBrowser: {
          url: 'https://example.com/cdn/loading-module/1.0.0/loading-module.legacy.browser.js',
          integrity: '344',
        },
      })
    );
    expect(mockElement.addEventListener.mock.calls[1][0]).toBe('load');
    expect(mockElement.addEventListener.mock.calls[1][1]()).toBeUndefined();
    expect(mockElement.src).toBe('https://example.com/cdn/loading-module/1.0.0/loading-module.browser.js');
    expect(new URL(mockElement.src).search).toBe('');
  });
});
