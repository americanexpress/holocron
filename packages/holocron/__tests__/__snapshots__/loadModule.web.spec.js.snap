// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadModule.web should reject with a helpful error when module data is not an object 1`] = `[TypeError: loadModule: moduleData must be an object]`;

exports[`loadModule.web should reject with a helpful error when module name is not a string 1`] = `[TypeError: loadModule: moduleName must be a string]`;
